src/object/srv_obj.c

```diff
@@ -720,85 +1046,220 @@ ds_obj_rw_handler(crt_rpc_t *rpc)
+		/* XXX: For new leader, even though the local replica
+		 *	has ever been modified before, but it doesn't
+		 *	know whether other replicas have also done the
+		 *	modification or not, so still need to dispatch
+		 *	the RPC to other replicas.
+		 */
+
+		D_ASSERT(orw->orw_dti_cos.ca_count == 0);
+
+		count = obj_prepare_cos(cont, cont_hdl, &orw->orw_oid,
+					&orw->orw_dkey, DCLT_PUNCH, &dti);
+		if (count < 0)
+			D_GOTO(reply, rc = count);
+
 		rc = ds_obj_req_disp_prepare(rpc->cr_opc,
 			orw->orw_shard_tgts.ca_arrays,
-			orw->orw_shard_tgts.ca_count,
-			obj_update_prefw, orw, obj_update_postfw, orw,
-			&obj_arg);
+			orw->orw_shard_tgts.ca_count, obj_update_prefw,
+			obj_update_postfw, orw, count, dti, &obj_arg);
 		if (rc != 0) {
 			D_ERROR(DF_UOID": ds_obj_req_disp_prepare failed %d.\n",
 				DP_UOID(orw->orw_oid), rc);
-			D_GOTO(out, rc);
+			D_GOTO(reply, rc);
 		}
 		D_ASSERT(obj_arg != NULL);
 
+force_resend:
 		rc = dss_ult_create(ds_obj_req_dispatch, obj_arg,
 				    DSS_ULT_IOFW, tag, 0, NULL);
 		if (rc != 0) {
-			D_ERROR(DF_UOID": ds_obj_update_dispatch failed %d.\n",
+			D_ERROR(DF_UOID": ds_obj_req_dispatch failed %d.\n",
 				DP_UOID(orw->orw_oid), rc);
 			ds_obj_req_disp_arg_free(obj_arg);
-			D_GOTO(out, rc);
+			D_GOTO(reply, rc);
 		}
+	} else {
+		count = orw->orw_dti_cos.ca_count;
+		dti = orw->orw_dti_cos.ca_arrays;
 	}
 
-	/* local RPC handler */
-	rc = ds_obj_rw_local_hdlr(rpc, tag, cont_hdl, cont, &ioh, update);
-	if (rc != 0)
-		D_ERROR(DF_UOID": ds_obj_rw_local_hdlr failed %d.\n",
-			DP_UOID(orw->orw_oid), rc);
+	if (!resend) {
+		if (srv_enable_dtx)
+			rc = vos_dtx_begin(&orw->orw_dti, &orw->orw_oid,
+				&orw->orw_dkey, cont->sc_hdl, orw->orw_epoch,
+				orw->orw_map_ver, DAOS_INTENT_UPDATE,
+				(orw->orw_flags & ORF_DTX_AOC) ? DTX_F_AOC : 0,
+				&dth);
+		else
+			rc = 0;
+		if (rc >= 0) {
+			if (dth != NULL &&
+			    !(orw->orw_flags & ORF_FROM_LEADER)) {
+				dth->dth_leader = 1;
+				if (!dispatch)
+					dth->dth_non_rep = 1;
+			}
+
+local:
+			/* local RPC handler */
+			rc = ds_obj_rw_local_hdlr(rpc, tag, cont_hdl, cont,
+						  &ioh, dth, map_ver, count,
+						  dti, update);
+			if (rc != 0)
+				D_ERROR(DF_UOID
+					": obj_rw_local_hdlr failed %d.\n",
+					DP_UOID(orw->orw_oid), rc);
+		}
+	}
 
 	/* wait dispatched IO's completion when needed */
-	if (dispatch)
+	if (dispatch) {
 		dispatch_rc = ds_obj_req_disp_wait(obj_arg);
+		if (rc >= 0 && !(obj_arg->flags & ORF_DTX_AOC)) {
+			if (dispatch_rc == -DER_INPROGRESS) {
+				/* XXX: The local modification has been done,
+				 *	but remote replica failed because of
+				 *	some uncommitted DTX, then it should
+				 *	be the case that we have ever failed
+				 *	to abort such DTX on remote replica.
+				 *	Let's retry and force abort related
+				 *	conflict (uncommitted) DTX by force.
+				 */
+
+				rpc_dispatched = false;
+				resend = true;
+				obj_arg->flags |= ORF_RESEND | ORF_DTX_AOC;
+
+				goto force_resend;
+			}
```

Comment:
This obj rw handler is too complicate now. probably we should separate the RPC handler into 
i/o handler to handle rpc from client side to dispatch RPC.
target i/o handler to handle rpc from the leader

I think that would make it looks simpler and cleaner.

Reply:
OK, I will consider to optimize and clean the code next step.